airflow/models/dagrun.py

```diff
@@ -334,6 +334,34 @@ def find(
 
         return qry.order_by(DR.execution_date).all()
 
+    @staticmethod
+    @provide_session
+    def find_duplicate(
+        dag_id: str,
+        run_id: str,
+        execution_date: datetime,
+        session: Session = None,
+    ) -> Optional['DagRun']:
+        """
+        Returns a dag run if there is an existing run for a dag at a specific run_id and execution_date.
+
+        :param dag_id: the dag_id to find duplicates for
+        :type dag_id: str
+        :param run_id: defines the run id for this dag run
+        :type run_id: str
+        :param execution_date: the execution date
+        :type execution_date: datetime.datetime
+        :param session: database session
+        :type session: sqlalchemy.orm.session.Session
+        """
+        DR = DagRun
+
+        qry = session.query(DR)
+
+        qry = qry.filter(DR.dag_id == dag_id, or_(DR.run_id == run_id, DR.execution_date == execution_date))
+
+        return qry.first()
```

Comment:
```suggestion
        return (
            session.query(cls)
			.filter(
			    cls.dag_id == dag_id,
			    or_(cls.run_id == run_id, cls.execution_date == execution_date),
			)
			.one_or_none()
		)
```

This should always return at most one row since we have a unique constraint on the database, but if it somehow returns multiple ones, we want to be loud about it to raise awareness to the inconsistency.